Fluid/Unit.cs

```diff
@@ -20,7 +20,7 @@ abstract public class Unit
         public readonly SamplerState Sampling = new SamplerState();
         internal RenderTarget2D Temporary;
         public const SpriteSortMode Sorting = SpriteSortMode.Immediate;
-        public const SurfaceFormat Surface = SurfaceFormat.HdrBlendable;
+        public const SurfaceFormat Surface = SurfaceFormat.HalfVector4;
```

Comment:
I suspect this change as first candidacy for broken fluid
Because I remember `SurfaceFormat` was a really pain during development
I tried a lot of them

Rest places use `Vector4` or `HalfVector4` types
But this one had `HdrBlendable` for a reason (I don't know what, sadly)
I remember that tried to make it `Vector4` too, but it didn't worked. 

Perhaps, something was changed inside `MonoGame` from that time
Hence `HdrBlendable` stopped to work too

Reply:
thanks.  i changed it to that because there was an exception thown with HdrBlendable and the types used elsewhere were HalfVector 4..   i think running the monogame from source is the next steop, i hate blindly hacking.  dev studio allows debugging HLSL now.  Monogame is wicked but i think targeting only windows and xbox one is my goal for this.   a 512 x512 fluid at 60 fps is incredible.. with CPU i cant even do 64x64 .   But for hi res one needs alot of RAM for all the intermediate surfaces.  I looked like you were targetting android and that is why you mighht have pulled out fluid.  its is hopeless < 1G RAM like phones where then.

Reply:
looking the monogame recent 3.6 source, HdrBlendable is not supported in   monogame source it does not support the type  HdrBlendable in GetSize



     //   public const SurfaceFormat Surface = SurfaceFormat.HalfVector4;
        public const SurfaceFormat Surface = SurfaceFormat.HdrBlendable;
unit.cs 
 
also, Data has the matching type
the private readonly HalfVector4[] data;

        public void Process(RenderTarget2D texture)
        {
            Copy(texture);
            input.GetData(data);  <=== exception Value does not fallinto the expected range.. 

looking in the monogame source it does not support the type  HdrBlendable in GetSize



, it calls 		StackTrace	"   at Microsoft.Xna.Framework.Graphics.GraphicsExtensions.GetSize(SurfaceFormat surfaceFormat)\r\n   at Microsoft.Xna.Framework.Graphics.Texture2D.ValidateParams[T](Int32 level, Int32 arraySlice, Nullable`1 rect, T[] data, Int32 startIndex, Int32 elementCount, Rectangle& checkedRect)\r\n   at Microsoft.Xna.Framework.Graphics.Texture2D.GetData[T](T[] data)\r\n   at Fluid.Data.Process(RenderTarget2D texture) in C:\\Users\\damia\\Desktop\\Escape-OnlyWindowsVersion\\Fluid\\Data.cs:line 38"

Reply:
https://blogs.msdn.microsoft.com/shawnhar/2010/07/09/surfaceformat-hdrblendable/

more on this... https://github.com/labnation/MonoGame/blob/master/MonoGame.Framework/Windows8/SharpDXHelper.cs1

to build monogame.. i ran a protobuild and build it.. i havent tried debugging shader in MG though..

Reply:
i tried using nuGet for mG Gl and MG sharpDX.. but the same issue..

Reply:
> types used elsewhere were HalfVector 4

As I remember there were different types of data, each required different format
For example, `Velocity` is decoded by `Vector2` or `float2` in shaders
`Pressure` is a scalar thing so only single `float` was required
`Color` had `r,g,b` components thus `Vector3` here
But at some places different `colors` were blended together, with `alpha` channel it's `r,g,b,a` and `Vector4` was used
An `Obstacle` probably also had distinct format

Another important thing is `bits` per variable
In `.cs` code `float` has `32` bits, shaders probably too
`Pressure` and `Obstacle` could use `16` bits because they didn't require high precision

These were just assumptions and possible clues
Some details can be wrong since it was `4` years ago
I didn't work with `Grapshics` and `Shaders` from that time. I am happy already I can remember such details
You need to verify them by your own

Reply:
> i hate blindly hacking

I understand you so good, me too
But know that a big part of `fluid simulation` was made by heavy usage of `blind hacking` or similar tricks
Especially with `format` issues and interaction with game objects
I was hurried a lot that time, just tried to get something working as fast as possible
I really ashamed reading that code now

Reply:
```cs
case SurfaceFormat.HdrBlendable:
    // TODO: This needs to check the graphics device and
    // return the best hdr blendable format for the device.
    return SharpDX.DXGI.Format.R16G16B16A16_Float;
```

I realized just now the development was made with `NVidia GeForce 750 Ti` videocard
But a problem occured with it so I pulled out the videocard from my PC and use `Intel` integrated graphics now
Maybe it can be a reason for `HdrBlendable` crash

I suggest to verify if your `videocard` supports required format `Float 16 bits per channel`
As I know there are many restrictions with `float` format for `GPU`. Such restriction was a reason why I couldn't laucnh it on `Android`

Reply:
> dev studio allows debugging HLSL now however, the fx goes to xnb, it might be debuggable in a temp folder tho somehow match with fx file source.

`4` years ago I did not debug shaders, it wasn't very helpful. 
Maybe I just used a poor tool

But a super useful trick was just to decode a required info with color data
So `Velocity` is `float2` which I decoded with color format `float3` leaving last `blue` component `0` like that: `float3 (Velocty.X, Velocity.Y, 0.0)`. Hence any velocity will change `red` or `green` component of a pixel. Like that:
![image](https://user-images.githubusercontent.com/5301844/36726527-ba4920aa-1bc2-11e8-9f5e-144227beb638.png)

However the color `float3 (-1.0, -1.0, 0.0)` is invalid so I had to clamp velocity into `[0,1]` range
That's why empty `Velocity` field is gray `(0.5, 0.5, 0.5)` not black

The same trick with `Pressure` but it's scalar value so color will be: `float3 (Pressure, 0.0, 0.0)`
`Obstacles` data also was debugged in such way: all that `East, South, West, ...` are pretty colorful things and visualized well

Reply:
for this kind of thing,  instead of clamping, I suggest linear mapping the -1f to -1, in the simulation, to positive 0 to 1 is better.    For this range,  since -1 to 1 range is length 2.. 

divide by 2 and add 1/2..

that brings -1f to 0.f.   1f stays 1f, 0f is 0.5f.  resolution is the same and no info lost and RGB representation